File: ts/src/packages/dapr/src/parts/dapr-repository.ts
import { Result, ok, err, ResultAsync } from 'neverthrow';
import {
  Aggregate,
  IEvent,
  PartitionKeys,
  AggregateProjector,
  EventStoreError,
  SortableUniqueId,
  EmptyAggregatePayload
} from '@sekiban/core';
import type { IAggregateEventHandlerActor, SerializableEventDocument } from '../actors/interfaces.js';
import type { SekibanDomainTypes } from '@sekiban/core';

/**
 * Repository implementation for Dapr actors, bridging between AggregateActor and AggregateEventHandlerActor.
 * This is the Dapr equivalent of Orleans' OrleansRepository and C#'s DaprRepository.
 */
export class DaprRepository {
  private currentAggregate: Aggregate;

  constructor(
    private readonly eventHandlerActor: IAggregateEventHandlerActor,
    private readonly partitionKeys: PartitionKeys,
    private readonly projector: AggregateProjector<any>,
    private readonly domainTypes: SekibanDomainTypes,
    currentAggregate: Aggregate
  ) {
    this.currentAggregate = currentAggregate;
  }

  getAggregate(): Result<Aggregate, never> {
    return ok(this.currentAggregate);
  }

  async save(
    lastSortableUniqueId: string,
    newEvents: IEvent[]
  ): Promise<Result<IEvent[], SekibanError>> {
    if (!newEvents || newEvents.length === 0) {
      return ok([]);
    }

    try {
      // Convert events to serializable documents using domain types
      const eventDocuments: SerializableEventDocument[] = newEvents.map(event => {
        // Use domain types to serialize the event
        const eventDocument = this.domainTypes.eventTypes.serializeEvent(event);
        
        return {
          id: event.id.toString(),
          sortableUniqueId: event.id.toString(),
          payload: eventDocument.payload,
          eventType: eventDocument.eventType,
          aggregateId: eventDocument.aggregateId,
          partitionKeys: event.partitionKeys,
          version: eventDocument.version,
          createdAt: eventDocument.timestamp.toISOString(),
          metadata: eventDocument.metadata
        };
      });

      // Call the event handler actor to append events
      const response = await this.eventHandlerActor.appendEventsAsync(
        lastSortableUniqueId,
        eventDocuments
      );

      if (!response.isSuccess) {
        return err(new EventStoreError(response.error || 'Failed to append events'));
      }

      // Note: Unlike the previous implementation, we should NOT update currentAggregate here
      // The caller (AggregateActor) will handle the aggregate update via getProjectedAggregate
      console.log(`[DaprRepository.save] Events saved: ${newEvents.length}, current version: ${this.currentAggregate.version}`);

      return ok(newEvents);
    } catch (error) {
      return err(new EventStoreError(error instanceof Error ? error.message : 'Unknown error'));
    }
  }

  async load(): Promise<Result<Aggregate, SekibanError>> {
    try {
      // Get all events from the event handler
      const eventDocuments = await this.eventHandlerActor.getAllEventsAsync();

      // Convert documents back to events using domain types for proper deserialization
      const events: IEvent[] = [];
      for (const doc of eventDocuments) {
        const eventDocument = {
          id: doc.id,
          eventType: doc.eventType,
          aggregateId: doc.aggregateId,
          aggregateType: doc.partitionKeys.group || '',
          payload: doc.payload,
          metadata: doc.metadata || {},
          timestamp: new Date(doc.createdAt),
          version: doc.version
        };
        
        const deserializedResult = this.domainTypes.eventTypes.deserializeEvent(eventDocument);
        if (deserializedResult.isErr()) {
          throw deserializedResult.error;
        }
        events.push(deserializedResult.value);
      }

      // Start with empty aggregate
      let aggregate = Aggregate.emptyFromPartitionKeys(this.partitionKeys);

      // Project all events
      for (const event of events) {
        const projectResult = this.projector.project(aggregate, event);
        if (projectResult.isErr()) {
          return err(projectResult.error);
        }
        aggregate = projectResult.value;
      }

      // Update current aggregate
      this.currentAggregate = aggregate;

      return ok(aggregate);
    } catch (error) {
      return err(new EventStoreError(error instanceof Error ? error.message : 'Unknown error'));
    }
  }

  getProjectedAggregate(projectedEvents: IEvent[]): Result<Aggregate, SekibanError> {
    try {
      console.log(`[getProjectedAggregate] Current version: ${this.currentAggregate.version}, projecting ${projectedEvents.length} events`);

      // Project the events onto the current aggregate to get a new aggregate
      let aggregate = this.currentAggregate;
      for (const event of projectedEvents) {
        const projectResult = this.projector.project(aggregate, event);
        if (projectResult.isErr()) {
          return err(projectResult.error);
        }
        aggregate = projectResult.value;
      }

      console.log(`[getProjectedAggregate] After projection - version: ${aggregate.version}`);
      return ok(aggregate);
    } catch (error) {
      return err(new EventStoreError(error instanceof Error ? error.message : 'Unknown error'));
    }
  }
}

File: pure/src/Sekiban.Pure.Dapr/Parts/DaprRepository.cs
using Dapr;
using ResultBoxes;
using Sekiban.Pure.Aggregates;
using Sekiban.Pure.Dapr.Actors;
using Sekiban.Pure.Documents;
using Sekiban.Pure.Events;
using Sekiban.Pure.Projectors;
namespace Sekiban.Pure.Dapr.Parts;

/// <summary>
///     Repository implementation for Dapr actors, bridging between AggregateActor and AggregateEventHandlerActor.
///     This is the Dapr equivalent of Orleans' OrleansRepository.
/// </summary>
public class DaprRepository
{
    private readonly SekibanDomainTypes _domainTypes;
    private readonly IAggregateEventHandlerActor _eventHandlerActor;
    private readonly IEventTypes _eventTypes;
    private readonly PartitionKeys _partitionKeys;
    private readonly IAggregateProjector _projector;
    private Aggregate _currentAggregate;
    public DaprRepository(
        IAggregateEventHandlerActor eventHandlerActor,
        PartitionKeys partitionKeys,
        IAggregateProjector projector,
        IEventTypes eventTypes,
        Aggregate currentAggregate,
        SekibanDomainTypes domainTypes)
    {
        _eventHandlerActor = eventHandlerActor ?? throw new ArgumentNullException(nameof(eventHandlerActor));
        _partitionKeys = partitionKeys ?? throw new ArgumentNullException(nameof(partitionKeys));
        _projector = projector ?? throw new ArgumentNullException(nameof(projector));
        _eventTypes = eventTypes ?? throw new ArgumentNullException(nameof(eventTypes));
        _currentAggregate = currentAggregate ?? throw new ArgumentNullException(nameof(currentAggregate));
        _domainTypes = domainTypes;
    }

    public ResultBox<Aggregate> GetAggregate() => ResultBox<Aggregate>.FromValue(_currentAggregate);

    public async Task<ResultBox<List<IEvent>>> Save(string lastSortableUniqueId, List<IEvent> newEvents)
    {
        if (newEvents == null || newEvents.Count == 0)
        {
            return ResultBox<List<IEvent>>.FromValue(new List<IEvent>());
        }

        try
        {
            // Convert events to serializable documents
            var eventDocuments = new List<SerializableEventDocument>();
            foreach (var @event in newEvents)
            {
                var document = await SerializableEventDocument.CreateFromEventAsync(
                    @event,
                    _domainTypes.JsonSerializerOptions);
                eventDocuments.Add(document);
            }

            // Call the event handler actor to append events
            var response = await _eventHandlerActor.AppendEventsAsync(lastSortableUniqueId, eventDocuments);

            if (!response.IsSuccess)
            {
                return ResultBox<List<IEvent>>.FromException(new InvalidOperationException(response.ErrorMessage));
            }

            // Note: Unlike the previous implementation, we should NOT update _currentAggregate here
            // The caller (AggregateActor) will handle the aggregate update via GetProjectedAggregate
            Console.WriteLine(
                $"[DaprRepository.Save] Events saved: {newEvents.Count}, current version: {_currentAggregate.Version}");

            return ResultBox<List<IEvent>>.FromValue(newEvents);
        }
        catch (Exception ex)
        {
            return ResultBox<List<IEvent>>.FromException(ex);
        }
    }

    public async Task<ResultBox<Aggregate>> Load()
    {
        try
        {
            // Get all events from the event handler
            IReadOnlyList<SerializableEventDocument> eventDocuments;
            try
            {
                eventDocuments = await _eventHandlerActor.GetAllEventsAsync();
            }
            catch (DaprApiException ex) when (ex.Message.Contains("500"))
            {
                // Actor doesn't exist yet, return empty list
                eventDocuments = new List<SerializableEventDocument>();
            }

            // Convert documents back to events
            var events = new List<IEvent>();
            foreach (var document in eventDocuments)
            {
                var eventResult = await document.ToEventAsync(_domainTypes);
                if (eventResult.HasValue && eventResult.Value is not null)
                {
                    events.Add(eventResult.Value);
                }
            }

            // Start with empty aggregate
            var aggregate = Aggregate.EmptyFromPartitionKeys(_partitionKeys);

            // Project all events
            aggregate = aggregate.Project(events, _projector).UnwrapBox();

            // Update current aggregate
            _currentAggregate = aggregate;

            return ResultBox<Aggregate>.FromValue(aggregate);
        }
        catch (Exception ex)
        {
            return ResultBox<Aggregate>.FromException(ex);
        }
    }

    public ResultBox<Aggregate> GetProjectedAggregate(List<IEvent> projectedEvents)
    {
        try
        {
            Console.WriteLine(
                $"[GetProjectedAggregate] Current version: {_currentAggregate.Version}, projecting {projectedEvents.Count} events");

            // Project the events onto the current aggregate to get a new aggregate
            var aggregate = _currentAggregate.Project(projectedEvents, _projector).UnwrapBox();

            Console.WriteLine($"[GetProjectedAggregate] After projection - version: {aggregate.Version}");
            return ResultBox<Aggregate>.FromValue(aggregate);
        }
        catch (Exception ex)
        {
            return ResultBox<Aggregate>.FromException(ex);
        }
    }
}
